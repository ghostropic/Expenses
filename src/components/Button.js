import React from 'react'
import styled from 'styled-components'

const Button = ({ text, onClick, disabled, className, type }) => (
  // eslint-disable-next-line react/button-has-type
  <button
    type={type}
    disabled={disabled}
    onClick={onClick}
    className={className}
  >
    {text}
  </button>
)

const StyledButton = styled(Button)`
  cursor: pointer;
  height: 48px;
  display: inline-block;
  padding: 0.5rem 0;
  width: 11rem;
  color: ${props => props.color || 'white'};
  border: 2px solid ${props => props.color || 'white'};
  background: ${props => props.background || 'transparent'};

  &.formButton {
    margin: 7px;
  }
  
  &.removeButton {
    background: palevioletred;
  }
 `

export default StyledButton
