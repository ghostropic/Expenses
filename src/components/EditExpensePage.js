import React from 'react'
import styled from 'styled-components'

const EditExpensePage = (props) => {
  console.log(props)
  return (
    <div>
      edit expenses for {props.match.params.id}
    </div>
  )
}

export default EditExpensePage