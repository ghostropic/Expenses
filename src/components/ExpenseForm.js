import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

import 'react-dates/initialize'
import 'react-dates/lib/css/_datepicker.css'

export default class ExpenseForm extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100) : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: ''
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({createdAt}))
    }
  }

  onFocusChange = ({focused}) => {
    this.setState(() => ({focused}))
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({description}))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount})) 
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({note}))
  }

  OnSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: 'this is an error message'}))
      console.log('error')
    } else {
      this.setState(() => ({error: ''}))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }

  render() {
    return (
    <div> 
      <p>Expense Form:</p>
      <div>{this.state.error}</div>
      <form onSubmit={this.OnSubmit}>
        <input 
        value={this.state.description} 
        onChange={this.onDescriptionChange} 
        type="text" 
        placeholder="description" 
        autoFocus
        />
        <input 
        type="text" 
        placeholder="amount"
        value={this.state.amount}
        onChange={this.onAmountChange} 
        /> 
        <textarea 
        placeholder="Add a note to your expense."
        value={this.state.note}
        onChange={this.onNoteChange}
        ></textarea> 
        <SingleDatePicker
          date={this.state.createdAt} // momentPropTypes.momentObj or null
          onDateChange={this.onDateChange} // PropTypes.func.isRequired
          focused={this.state.focused} // PropTypes.bool
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
          id="expense_form_calender" // PropTypes.string.isRequired,
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <button>Add Expense</button>
      </form>
    </div>
    )
  }
}
