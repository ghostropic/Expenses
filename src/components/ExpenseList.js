import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

export const ExpenseList = ({ expenses }) => (
  <div>
    <p>{`Total: ${selectExpensesTotal(expenses)}`}</p>
    <hr />
    {
      expenses.length === 0
        ? (<p>No Expense to list. Try clearing the date range.</p>)
        : (expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />))
    }
  </div>
)

const MapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(MapStateToProps)(ExpenseList)
