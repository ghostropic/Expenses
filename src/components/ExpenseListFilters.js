import React from 'react'
import {connect} from 'react-redux'
import * as actionFilters from '../actions/filters'

const ExpenseListFilters = (props) => (
  <div>
    <span>Filter:</span>
    <input type="text" value={props.filters.text} onChange={(e) => {
      props.dispatch(actionFilters.setTextFilter(e.target.value))
    }}/>
    <select value={props.filters.sortBy} onChange={(e) => {
      e.target.value === 'date' ?
        props.dispatch(actionFilters.sortByDate()) :
        props.dispatch(actionFilters.sortByAmount())
    }}>
      <option value="amount">Amount</option>
      <option value="date">Date</option>
    </select>
  </div>
)

const MapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}

export default connect(MapStateToProps)(ExpenseListFilters) 