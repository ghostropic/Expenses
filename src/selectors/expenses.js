import moment from 'moment'

// getVisibleExpenses - passes expense and filters and determines what should be send along the expense list

export default (expenses, {text, sortBy, startDate, endDate}) => { // two arguments: expenses and destructured filters
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
    // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDates
    const textMatch = text === '' || expense.description.toLowerCase().includes(text.toLowerCase())
    return startDateMatch && endDateMatch && textMatch
  }).sort((a,b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  })
  // return [expenses, startDate]
}