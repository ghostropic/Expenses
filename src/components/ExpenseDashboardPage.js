import React from 'react'
import styled from 'styled-components'

import ExpenseList from './ExpenseList';

const ExpenseDashboardPage = () => (
  <div>
    <div>dashboard</div>
    <ExpenseList/>
  </div>
)

// const ExpenseDashboardPage = () => {
//   return (
//     <div>
//       dashboard
//     </div>
//   )
// }

export default ExpenseDashboardPage