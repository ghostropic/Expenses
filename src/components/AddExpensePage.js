import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import ExpenseForm from './ExpenseForm'
import expenses from '../reducers/expenses'
import { addExpense } from '../actions/expenses'

const AddExpensePage = (props) => (
  <div>
    add expense:
    <ExpenseForm
      onSubmit={(expense) => {
        props.dispatch(addExpense(expense))
        props.history.push('/')
      }}
    />
  </div>
)

export default connect()(AddExpensePage)