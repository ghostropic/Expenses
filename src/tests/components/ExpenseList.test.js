import React from 'react'
import { shallow, mount } from 'enzyme'

import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses'

// jest.mock('../../../src/components/ExpenseListItem', () => 
//   () => {}
// )

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />)

  expect(wrapper).toMatchSnapshot()
})


test('should render ExpenseList with empty message', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)

  expect(wrapper).toMatchSnapshot()
})
