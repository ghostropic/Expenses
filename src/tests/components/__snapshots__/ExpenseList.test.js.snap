// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    Total: $0.00
  </p>
  <hr />
  <p>
    No Expense to list. Try clearing the date range.
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <p>
    Total: $601.95
  </p>
  <hr />
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="gum"
    id="1"
    key="1"
    note="one pack of bubble gum"
  />
  <ExpenseListItem
    amount={15000}
    createdAt={-345600000}
    description="rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={45000}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
